site/_posts/2024-11-27-fish-4.md

```diff
@@ -0,0 +1,78 @@
+
+fish-shell is approaching its 4.0 milestone, the first release since porting from C++ to Rust. Today we are announcing an open beta, inviting all users to try out the upcoming 4.0 release.
+
+### **Try the fish-shell 4.0 beta at [fishshell.com](https://fishshell.com)**. View the [release notes here](https://fishshell.com/docs/current/relnotes.html).
+
+If you are an existing fish user, or are curious to give it a try, this is a great opportunity to improve your command line experience while contributing to fish. Please report any issues you encounter by filing a [GitHub Issue](http://github.com/fish-shell/fish-shell/issues), or on our [Matrix channel](https://matrix.to/#/#fish-shell:matrix.org), which is also a great place to ask questions. Newbie questions welcome!
+
+The remainder of this post will discuss the mechanics of the port.
+
+### By the numbers
+
+The port was proposed in February 2023. January 2024 marked the elimination of C++ from the fish-shell codebase; this was then followed by a period of testing, fixes, and preparation for the release.
+
+Prior to the port, fish-shell had over 47k lines of code in C++ source files, and nearly 8k lines in C++ header files. Today, fish has over 75k lines of Rust. This increase is explained by Rust's explicit `use` constructs compared to C++'s umbrella headers, `rustfmt` vs `clang-format` differences, the need to re-implement `printf`, and some additional features added in the interim.
+
+### Why port
+
+fish-shell was originally written in C, in 2005. It was ported to C++ in 2011, for improved memory management, and C++ served fish-shell well for a decade.
+
+But eventually, problems arose. Modern C++ is challenging to learn and makes it difficult to attract new developers. fish-shell maintains compatibility with older operating systems, limiting us to C++11 with no clear upgrade path. Lastly, fish-shell has a long-term goal of enabling fully multithreaded execution, and prototypes of this in C++ proved unpromising. For example, it was too easy to accidentally share objects across threads, with only post-hoc tools like Thread Sanitizer to prevent it.
+
+So we decided to port to a new language.
+
+### How to port
+
+A key requirement was that the port would be _incremental_: not a "big bang" rewrite, but instead one module or even function at a time, with a usable shell and all tests passing at each commit. This required strong language interoperability (i.e. an FFI), as both C++ and the target language would coexist for a period.
+
+We maintained the existing style in the ported code. This resulted in code that was not exactly idiomatic, but was easier to compare and review. For example, ported code used for loops instead of iterator chains. After we had confidence in the correctness of the ported code, we refactored it to a more idiomatic style.
```

Comment:
It's probably worth it to note that this is *ongoing* - we still have some clunky spots that could do with cleaning up.

A good example for this is errno - https://github.com/fish-shell/fish-shell/commit/e1d539c7b6268eca2c38e83a7a849912ff489620 replaced code that *set* errno to signal an error with a real error type.